### PR TITLE
Add support for resolving .pom files to the maven resolver.

### PR DESCRIPTION
**Pull request by mikekap**
Also adds support for multiple repo urls & speeds up downloads by parallelizing.

### CODE REVIEW

File: src/com/facebook/buck/maven/BuckDepComparator.java [hunk 1]
```diff
@@ -27,6 +27,16 @@
...
+    if ("PUBLIC".equals(left) && "PUBLIC".equals(right)) {
+      return 0;
+    }
+    if ("PUBLIC".equals(left)) {
+      return -1;
+    }
+    if ("PUBLIC".equals(right)) {
+      return 1;
+    }
+
```
Comment (beefon) [inline src/com/facebook/buck/maven/BuckDepComparator.java]: I think its better to move "PUBLIC" to a constant value.

File: src/com/facebook/buck/maven/Resolver.java [hunk 2]
```diff
@@ -92,73 +114,133 @@
   private final Path buckRepoRoot;
   private final Path buckThirdPartyRelativePath;
   private final LocalRepository localRepo;
-  private final ImmutableList<RemoteRepository> repos;
   private final ServiceLocator locator;
   private final VersionScheme versionScheme = new GenericVersionScheme();
+  private final ModelBuilder modelBuilder;
+
+  private ImmutableList<Dependency> initialRequest;
```
Comment (beefon) [inline src/com/facebook/buck/maven/Resolver.java]: Maybe its better to rename this ivar. `initialRequest` and `List` types don't sound together well. Maybe something like `initialRequestDeps` or something like that?

File: src/com/facebook/buck/maven/Resolver.java [hunk 3]
```diff
@@ -92,73 +114,133 @@
...
+    repos = ImmutableList.<RemoteRepository>builder()
+        .addAll(repos)
+        .addAll(getReposFromPom(model))
+        .build();
   }
 
-  public void resolve(String... mavenCoords) throws RepositoryException, IOException {
-    RepositorySystem repoSys = locator.getService(RepositorySystem.class);
-    RepositorySystemSession session = newSession(repoSys);
+  public void resolve() throws Exception {
```
Comment (beefon) [inline src/com/facebook/buck/maven/Resolver.java]: Is it necessary to generalise the type of exception? Or some callable method throws `Exception`?

Comment (beefon) [inline src/com/facebook/buck/maven/Resolver.java]: Also, this method looks quite big. Maybe splitting it to smaller ones can help a bit.

File: src/com/facebook/buck/maven/Resolver.java [hunk 4]
```diff
@@ -172,13 +254,19 @@ private void downloadArtifact(
     }
 
     // Populate visibility
-    Iterable<Artifact> outgoing = graph.getOutgoingNodesFor(artifactToDownload);
-    for (Artifact artifact : outgoing) {
-      String groupName = getProjectName(artifact);
-      if (!groupName.equals(projectName)) {
-        library.addVisibility(buckThirdPartyRelativePath, artifact);
+    if (initialArtifacts.contains(buildKey(artifactToDownload))) {
+      library.addVisibility("PUBLIC");
```
Comment (beefon) [inline src/com/facebook/buck/maven/Resolver.java]: You can reuse const for PUBLIC here then.

File: src/com/facebook/buck/maven/Resolver.java [hunk 5]
```diff
@@ -361,9 +450,9 @@ private Path copy(ArtifactResult result, Path destDir) throws IOException {
         }
 
         // TODO(shs96c): Do we always want optional dependencies?
-//        if (dependency.isOptional()) {
-//          continue;
-//        }
+        // if (dependency.isOptional()) {
+        //   continue;
+        // }
```
Comment (beefon) [inline src/com/facebook/buck/maven/Resolver.java]: Can you revert this change to make PR a bit lighter please?

File: src/com/facebook/buck/maven/Resolver.java [hunk 6]
```diff
@@ -439,25 +584,38 @@ private String buildKey(Artifact artifact) {
...
-        m2,
-        mavenCentral)
-        .resolve(coords);
+        m2);
+
+    if (args.length == 4) {
+      resolver.loadPomFile(Paths.get(args[3]));
+    } else {
+      List<String> repos = new ArrayList<>();
+      int i;
```
Comment (beefon) [inline src/com/facebook/buck/maven/Resolver.java]: `i` is used outside `for` loop, so it's better to name it accordingly. Like `indexOfShellPassThroughArgListSeparator` or how ever you'd call this special argument.

**Comment by beefon** [comment]
And please make sure CI builds are all green.

**Comment by mikekap** [comment]
I'm going to pause working on this until #691 is closed. It's a pretty big merge conflict at this point. Also CI is busted for every PR so I can't make it green.

**Comment by facebook-github-bot-7** [comment]
@mikekap updated the pull request.

**Comment by mikekap** [comment]
Updated and addressed comments.

File: src/com/facebook/buck/maven/Resolver.java [hunk 7]
```diff
@@ -98,75 +119,116 @@
...
-      List<String> extraVisibility = specifiedArtifacts.contains(artifact.toString()) ?
-        this.visibility : ImmutableList.<String>of();
-      downloadArtifact(artifact, repoSys, session, buckFiles, graph, extraVisibility);
-    }
+  private ImmutableSetMultimap<Path, Prebuilt> downloadArtifacts(
+      final RepositorySystem repoSys,
+      final RepositorySystemSession session,
+      final MutableDirectedGraph<Artifact> graph) throws ExecutionException, InterruptedException {
+    ListeningExecutorService exec = MoreExecutors.listeningDecorator(
+        Executors.newWorkStealingPool());
```
Comment (marcinkosiba) [inline src/com/facebook/buck/maven/Resolver.java]: to make it easier to debug we tend to use the MoreExecutors wrappers which give the threads in the pool useful names. Any chance you could update this to name the threads something like "MavenArtifactDownload"?

File: src/com/facebook/buck/maven/Resolver.java [hunk 8]
```diff
@@ -98,75 +119,116 @@
...
+            .transform(new Function<Artifact, Callable<Map.Entry<Path, Prebuilt>>>() {
+              @Override
+              public Callable<Map.Entry<Path, Prebuilt>> apply(final Artifact artifact) {
+                return new Callable<Map.Entry<Path, Prebuilt>>() {
+                  @Override
+                  public Map.Entry<Path, Prebuilt> call() throws Exception {
+                    try {
+                      return downloadArtifact(artifact, repoSys, session, graph);
+                    } catch (Throwable t) {
+                      throw new RuntimeException(t);
```
Comment (marcinkosiba) [inline src/com/facebook/buck/maven/Resolver.java]: why do you need to do this?

File: src/com/facebook/buck/maven/Resolver.java [hunk 9]
```diff
@@ -98,75 +120,114 @@
...
+      final RepositorySystemSession session,
+      final MutableDirectedGraph<Artifact> graph) throws ExecutionException, InterruptedException {
+    ListeningExecutorService exec = MoreExecutors.listeningDecorator(
+        Executors.newFixedThreadPool(
+            Runtime.getRuntime().availableProcessors(),
+            new MostExecutors.NamedThreadFactory("artifact download")));
+
+    @SuppressWarnings("unchecked")
+    List<ListenableFuture<Map.Entry<Path, Prebuilt>>> results =
+        (List) exec.invokeAll(FluentIterable.from(graph.getNodes())
```
Comment (Coneko) [inline src/com/facebook/buck/maven/Resolver.java]: ant is complaining that you're casting to a parametrised type without specifying the parameter.

`@SuppressWarnings("unchecked")` will let you cast to any parameter, but you still have to provide one.

Still, I don't understand how this is supposed to work: you're passing a list of callables instead of a list of futures?

**Comment by Coneko** [comment]
Do you get notified on inline comments now? I remember we used to have to ping with emojis back in my day.

File: src/com/facebook/buck/maven/Resolver.java (hunk 9 above)
Comment (mikekap) [inline src/com/facebook/buck/maven/Resolver.java]: Added gross casting to please javac.

I'm just using `invokeAll` on an `Iterable<Callable<...>>`. The `listeningDecorator` returns `ListenableFuture`s, but the interface demands `List<Future<...>>`, so hence the casting.

**Comment by mikekap** [comment]
Yep, I get notified for inline comments. Emojis are always welcome though :)

File: src/com/facebook/buck/maven/Resolver.java [hunk 10]
```diff
@@ -98,75 +120,115 @@
...
+                };
+              }
+            })
+            .toList());
 
-    createBuckFiles(buckFiles);
+    try {
+      return ImmutableSetMultimap.<Path, Prebuilt>builder()
+          .orderValuesBy(Ordering.natural())
+          .putAll(Futures.allAsList(results).get())
```
Comment (Coneko) [inline src/com/facebook/buck/maven/Resolver.java]: Sorry I still don't understand how this works: why can you call `Futures.allAsList` on a list of callables?

Comment (mikekap) [inline src/com/facebook/buck/maven/Resolver.java]: I think you're misreading the code a bit. 

The code looks like:
```java
List<ListenableFuture<...>> results = exec.invokeAll(List<Callable<...>>);
```

`invokeAll` is a `List<Callable<...>> => List<ListenableFuture<...>>` and `Futures.allAsList` just transforms `List<ListenableFuture<...>> => ListenableFuture<List<...>>` so I could call `.get`. This is all just a very annoying way of doing a parallel map - I'm more than happy to change it if there's a better solution.

Comment (Coneko) [inline src/com/facebook/buck/maven/Resolver.java]: Oh I see, sorry I must have been blind.

**Comment by Coneko** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-1** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/913220645463410/int_phab) to review.

**Comment by Coneko** [comment]
I don't get what is going on, I'm getting this error on CI internally:

```
  [javadoc] Constructing Javadoc information...
  [javadoc] /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/third-party/java/maven/maven-model-builder-3.2.5.jar(org/apache/maven/model/composition/DefaultDependencyManagementImporter.class): warning: Cannot find annotation method 'role()' in type 'Component': class file for org.codehaus.plexus.component.annotations.Component not found
  [javadoc] /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/third-party/java/maven/maven-model-builder-3.2.5.jar(org/apache/maven/model/management/DefaultDependencyManagementInjector.class): warning: Cannot find annotation method 'role()' in type 'Component'
  [javadoc] /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/third-party/java/maven/maven-model-builder-3.2.5.jar(org/apache/maven/model/management/DefaultPluginManagementInjector.class): warning: Cannot find annotation method 'role()' in type 'Component'
  [javadoc] /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/third-party/java/maven/maven-model-builder-3.2.5.jar(org/apache/maven/model/plugin/DefaultPluginConfigurationExpander.class): warning: Cannot find annotation method 'role()' in type 'Component'
  [javadoc] /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/third-party/java/maven/maven-model-builder-3.2.5.jar(org/apache/maven/model/profile/DefaultProfileSelector.class): warning: Cannot find annotation method 'role()' in type 'Component'
  [javadoc] /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/third-party/java/maven/maven-model-builder-3.2.5.jar(org/apache/maven/model/profile/DefaultProfileSelector.class): warning: Cannot find annotation method 'role()' in type 'Requirement': class file for org.codehaus.plexus.component.annotations.Requirement not found
  [javadoc] Standard Doclet version 1.7.0_45
  [javadoc] Building tree for all the packages and classes...
  [javadoc] Building index for all the packages and classes...
  [javadoc] Building index for all classes...
  [javadoc] Generating /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/build/javadoc-all/help-doc.html...
  [javadoc] 6 warnings
     [exec] Unexpected Javadoc errors (7):
     [exec]   [javadoc] /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/third-party/java/maven/maven-model-builder-3.2.5.jar(org/apache/maven/model/composition/DefaultDependencyManagementImporter.class): warning: Cannot find annotation method 'role()' in type 'Component': class file for org.codehaus.plexus.component.annotations.Component not found
     [exec]   [javadoc] /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/third-party/java/maven/maven-model-builder-3.2.5.jar(org/apache/maven/model/management/DefaultDependencyManagementInjector.class): warning: Cannot find annotation method 'role()' in type 'Component'
     [exec]   [javadoc] /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/third-party/java/maven/maven-model-builder-3.2.5.jar(org/apache/maven/model/management/DefaultPluginManagementInjector.class): warning: Cannot find annotation method 'role()' in type 'Component'
     [exec]   [javadoc] /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/third-party/java/maven/maven-model-builder-3.2.5.jar(org/apache/maven/model/plugin/DefaultPluginConfigurationExpander.class): warning: Cannot find annotation method 'role()' in type 'Component'
     [exec]   [javadoc] /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/third-party/java/maven/maven-model-builder-3.2.5.jar(org/apache/maven/model/profile/DefaultProfileSelector.class): warning: Cannot find annotation method 'role()' in type 'Component'
     [exec]   [javadoc] /data/sandcastle/boxes/instance-buck-git-buck-verification-linux/third-party/java/maven/maven-model-builder-3.2.5.jar(org/apache/maven/model/profile/DefaultProfileSelector.class): warning: Cannot find annotation method 'role()' in type 'Requirement': class file for org.codehaus.plexus.component.annotations.Requirement not found
     [exec]   [javadoc] 6 warnings
```

Running `ant verify-javadoc` but I cannot repro it locally. Any ideas?

File: src/com/facebook/buck/maven/Resolver.java [hunk 11]
```diff
@@ -404,19 +470,67 @@ private boolean isTestTime(Dependency dependency) {
...
+        .transform(new Function<org.apache.maven.model.Dependency, Dependency>() {
+          @Override
+          public Dependency apply(org.apache.maven.model.Dependency dep) {
+            DefaultArtifact artifact = new DefaultArtifact(
+                String.format("%s:%s:%s", dep.getGroupId(), dep.getArtifactId(), dep.getVersion()));
+            ImmutableList<Exclusion> exclusions = FluentIterable.from(dep.getExclusions())
+                .transform(new Function<org.apache.maven.model.Exclusion, Exclusion>() {
+                  @Override
+                  public Exclusion apply(org.apache.maven.model.Exclusion input) {
+                    return new Exclusion(input.getGroupId(), input.getArtifactId(), "", "jar");
```
Comment (kageiit) [inline src/com/facebook/buck/maven/Resolver.java]: what about the packaging type?

if you hardcode jars, android aar poms will not be resolved correctly

Comment (kageiit) [inline src/com/facebook/buck/maven/Resolver.java]: I think this can be generic to include both the classifier and packaging type as well

**Comment by mikekap** [comment]
@Coneko I managed to repro this locally and added a fix. If you install that particular revision of java (i.e. not just the latest java7u80, but java7u45), you'll see that failure. It's not terribly easy to install it either since the installer doesn't work on the latest osx.

The fix is just to add the annotations to the classpath - they were indeed missing. I added the fix as a separate commit since I'm not quite sure if I'm doing that right. Either way it's definitely strange that this happens on a _particular_ version of java.

File: src/com/facebook/buck/maven/Resolver.java (hunk 11 above)
Comment (mikekap) [inline src/com/facebook/buck/maven/Resolver.java]: These are exclusions - the maven model doesn't have classifiers or packaging types for exclusions. Though it looks like this was indeed wrong - I changed the logic around here to better match what other places are doing (using "*" "*")
